src/contracts/ARMRouter.sol

```diff
@@ -0,0 +1,408 @@
+        require(amounts[0] <= msg.value, "ARMRouter: EXCESSIVE_INPUT");
+
+        // Wrap ETH to WETH
+        WETH.deposit{value: amounts[0]}();
+
+        // Perform the swaps along the path
+        _swapsForExactTokens(amounts, path, to);
+
+        // Refund any excess ETH to the sender
+        if (msg.value > amounts[0]) payable(msg.sender).transfer(msg.value - amounts[0]);
+    }
+
+    /// @notice Swaps as few input tokens as possible to receive an exact amount of ETH, along the route determined by the path.
+    /// @param amountOut The exact amount of ETH to receive.
+    /// @param amountInMax The maximum amount of input tokens that can be used for the swap.
+    /// @param path An array of token addresses representing the swap path.
+    /// @param to The address that will receive the ETH.
+    /// @param deadline The timestamp by which the transaction must be completed.
+    /// @return amounts An array of token amounts for each step in the swap path.
+    function swapTokensForExactETH(
+        uint256 amountOut,
+        uint256 amountInMax,
+        address[] calldata path,
+        address to,
+        uint256 deadline
+    ) external ensure(deadline) returns (uint256[] memory amounts) {
+        // Ensure the last token in the path is WETH
+        require(path[path.length - 1] == address(WETH), "ARMRouter: INVALID_PATH");
+
+        // Calculate the required input amounts for the desired output
+        amounts = _getAmountsIn(amountOut, path);
+
+        // Ensure the required input does not exceed the maximum allowed
+        require(amounts[0] <= amountInMax, "ARMRouter: EXCESSIVE_INPUT");
+
+        // Transfer the input tokens from the sender to this contract
+        IERC20(path[0]).transferFrom(msg.sender, address(this), amounts[0]);
+
+        // Perform the swaps along the path
+        _swapsForExactTokens(amounts, path, address(this));
+
+        // Unwrap WETH to ETH and transfer to the recipient
+        WETH.withdraw(amounts[amounts.length - 1]);
+        payable(to).transfer(amounts[amounts.length - 1]);
+    }
+
+    ////////////////////////////////////////////////////
+    ///                 Internal Logic
+    ////////////////////////////////////////////////////
+    /// @notice Internal function to perform swaps along the specified path.
+    /// @param amountIn The amount of input tokens to swap.
+    /// @param path The swap path as an array of token addresses.
+    /// @param to The address that will receive the output tokens.
+    /// @return amounts An array of token amounts for each step in the swap path.
+    function _swapExactTokenFor(uint256 amountIn, address[] memory path, address to)
+        internal
+        returns (uint256[] memory amounts)
+    {
+        // Initialize the amounts array
+        amounts = new uint256[](path.length);
+        amounts[0] = amountIn;
+
+        // Perform the swaps along the path
+        uint256 len = path.length;
```

Comment:
Are you sure this gives a gas improvement? It does for sure if path is a storage variable but I've found there was no improvement when getting the length of a memory variable